### PR TITLE
Add missing docstrings in print.jl

### PR DESCRIPTION
**Pull request by odow**
Part of https://github.com/jump-dev/JuMP.jl/issues/1825

Also related to https://github.com/jump-dev/JuMP.jl/pull/2449, which adds some other missing docstrings.

### CODE REVIEW

File: src/print.jl [hunk 1]
```diff
@@ -634,13 +651,32 @@ function constraint_string(
...
+    constraint_string(
+        print_mode,
+        ref::ConstraintRef;
+        in_math_mode::Bool = false)
+
+Return a string representation of the constraint `ref`, given the `print_mode`.
+
+`print_mode` should be `IJuliaMode` or `REPLMode`.
+"""
 function constraint_string(print_mode, ref::ConstraintRef; in_math_mode = false)
```
Comment (odow) [inline src/print.jl:665]: Should these even be public methods? There are `Base.show(io::IO, ref::ConstraintRef)` and `Base.show(io::IO, ::MIME"text/latex", ref::ConstraintRef)`, which we should encourage the use-of instead.

File: src/print.jl [hunk 2]
```diff
@@ -634,13 +651,32 @@ function constraint_string(
...
 #------------------------------------------------------------------------
+
+"""
+    nl_expr_string(model::Model, print_mode, c::_NonlinearExprData)
+
+Return a string representation of the nonlinear expression `c` belonging to
+`model`, given the `print_mode`. `print_mode` should be `IJuliaMode` or
+`REPLMode`.
+"""
 function nl_expr_string(model::Model, print_mode, c::_NonlinearExprData)
```
Comment (odow) [inline src/print.jl:680]: Ditto on whether this should be exported.

File: src/print.jl [hunk 3]
```diff
@@ -682,6 +718,13 @@ end
...
 # TODO: Printing is inconsistent between regular constraints and nonlinear
 # constraints because nonlinear constraints don't have names.
+"""
+    nl_constraint_string(model::Model, mode, c::_NonlinearConstraint)
+
+Return a string representation of the nonlinear constraint `c` belonging to
+`model`, given the `print_mode`. `print_mode` should be `IJuliaMode` or
+`REPLMode`.
+"""
 function nl_constraint_string(model::Model, mode, c::_NonlinearConstraint)
```
Comment (odow) [inline src/print.jl:728]: Ditto on whether this should be exported.

File: src/print.jl [hunk 4]
```diff
@@ -23,8 +23,25 @@
...
+An abstract type used for dispatching printing.
+"""
 abstract type PrintMode end
+
+"""
+    REPLMode
+
+A type used for dispatching printing. Produces plain-text representations.
+"""
 abstract type REPLMode <: PrintMode end
```
Comment (odow) [inline src/print.jl:38]: A few questions on these. Why are the abstract? Is anyone subtyping them? Ideally, for JuMP 1.0 we would replace and use `MIME"text/plain"` and `MIME"text/latex"` instead.

File: src/print.jl (hunk 1 above)
Comment (blegat) [reply src/print.jl:665]: Yes, extension might add methods to this function

File: src/print.jl (hunk 2 above)
Comment (blegat) [reply src/print.jl:680]: We should probably revisit these questions once https://github.com/jump-dev/MathOptInterface.jl/pull/1180 is done. I'm waitinh

File: src/print.jl (hunk 1 above)
Comment (odow) [reply src/print.jl:665]: The could implement `show` directly though, and we could use that.

File: src/print.jl (hunk 2 above)
Comment (odow) [reply src/print.jl:680]: The printing PR is ready from my side: https://github.com/jump-dev/JuMP.jl/pull/2449

File: src/print.jl (hunk 1 above)
Comment (blegat) [reply src/print.jl:665]: I think the difference is that `is_math_mode` can be used to avoid `$$` to be added twice.

**Comment by odow** [comment]
I'll merge the docstrings for now. We can consider a bigger printing refactor later.

**Pull request merged**